### PR TITLE
initial draft of cjd-casc post

### PR DESCRIPTION
**Pull request by mdelgin**

### CODE REVIEW

**Comment by kmadel** [comment]
I just did a wick read - really good stuff @mdelgin - I may review tomorrow morning but hopefully someone else beats me to it!

**Comment by jhendrick** [comment]
Did a read through, looks great, nicely done @mdelgin.

File: content/posts/cjd-casc.md [hunk 1]
```diff
@@ -0,0 +1,548 @@
...
+  name: "Matt Elgin"
+title: "Self-Updating Jenkins: GitOps for Jenkins Configuration"
+date: 2019-07-01T17:00:00-04:00
+showDate: true
+draft: true
+tags: ["jenkins","cloudbees jenkins distribution","gitops","kubernetes","kaniko","docker","cert-manager"]
+---
+# Self-Updating Jenkins: GitOps for Jenkins Configuration
+
+In this blog post, we'll walk through creating a self-updating instance of [CloudBees Jenkins Distribution](https://www.cloudbees.com/products/cloudbees-jenkins-distribution) fully configured-as-code stored in a Git repository.
```
Comment (cvega) [inline content/posts/cjd-casc.md]: I got a little caught up on the last sentence when reading, could be a personal problem that only I'm experiencing. What you wrote makes total sense, I just had to reread it. 

`fully configured, as code, and stored in a git repository`

**Comment by cvega** [comment]
I had a chance to review. Looks good. I added a suggestion, but I had to grasp at straws to find something. Thanks for putting this together.

File: content/posts/cjd-casc.md [hunk 2]
```diff
@@ -0,0 +1,548 @@
...
+author:
+  name: "Matt Elgin"
+title: "Self-Updating Jenkins: GitOps for Jenkins Configuration"
+date: 2019-07-01T17:00:00-04:00
+showDate: true
+draft: true
+tags: ["jenkins","cloudbees jenkins distribution","gitops","kubernetes","kaniko","docker","cert-manager"]
+---
+
+In this blog post, we'll walk through creating a self-updating instance of [CloudBees Jenkins Distribution](https://www.cloudbees.com/products/cloudbees-jenkins-distribution) fully configured-as-code stored in a Git repository.
```
Comment (kmadel) [inline content/posts/cjd-casc.md]: perhaps add 'the' - so it reads "creating a self-updating instance of the [CloudBees Jenkins Distribution]"

File: content/posts/cjd-casc.md [hunk 3]
```diff
@@ -0,0 +1,548 @@
...
+# grant cluster-admin to user
+kubectl create clusterrolebinding cluster-admin-binding \ --clusterrole cluster-admin \ --user $(gcloud config get-value account)
+# deploy nginx ingress controller resources
+kubectl apply -f https://raw.githubusercontent.com/kubernetes/ingress-nginx/master/deploy/static/mandatory.yaml
+kubectl apply -f https://raw.githubusercontent.com/kubernetes/ingress-nginx/master/deploy/static/provider/cloud-generic.yaml
+```
+
+Next, let's look at the manifest file that will deploy the necessary resources for CJD using the [cjd.yaml](https://github.com/cb-technologists/cjd-casc/blob/master/cjd.yaml) manifest file.
+
+First, we create a ServiceAccount, a Role with the necessary permissions to manage agents and perform the required update actions, and a RoleBinding to connect the two.
```
Comment (kmadel) [inline content/posts/cjd-casc.md]: Perhaps format K8s objects as code - ie: "First, we create a `ServiceAccount`, a `Role`"

File: content/posts/cjd-casc.md [hunk 4]
```diff
@@ -0,0 +1,548 @@
...
+  apiGroup: rbac.authorization.k8s.io
+  kind: Role
+  name: cjd
+  # namespace: cjd
+subjects:
+- kind: ServiceAccount
+  name: cjd
+```
+
+Next, we create a Service that exposes ports for access to the web interface and for master-agent communication:
```
Comment (kmadel) [inline content/posts/cjd-casc.md]: "Next, we create a Service that exposes ports for access to the **CJD** web interface and for master-agent communication:"

File: content/posts/cjd-casc.md [hunk 5]
```diff
@@ -0,0 +1,548 @@
...
+    spec:
+      accessModes: [ "ReadWriteOnce" ]
+      resources:
+        requests:
+          storage: 5Gi
+```
+
+## Configuring with Jenkins Configuration-as-Code Plugin
+
+With the YAML for the required CJD Kubernetes resources laid out, we'll now go into the code handling the configuration of the master . While detailing the StatefulSet above, we mentioned that a custom Docker image is used for the CJD container. The Dockerfile for this image can be found below:
```
Comment (kmadel) [inline content/posts/cjd-casc.md]: "the master ." remove extra space after master

File: content/posts/cjd-casc.md [hunk 6]
```diff
@@ -0,0 +1,548 @@
...
+
+Once `cert-manager` has been deployed in its new namespace, we'll next need to deploy the Issuer to our `cjd` namespace. (**Note**: on initial setup of `cert-manager`, it's probably prudent to heed the Quick-Start's recommendation to create a staging Issuer first to minimize the risk of being rate limited by Let's Encrypt. For brevity, we'll only walk through the production Issuer creation here.)
```
Comment (kmadel) [inline content/posts/cjd-casc.md]: For the note use:
>**Note**: ...

File: content/posts/cjd-casc.md [hunk 7]
```diff
@@ -0,0 +1,548 @@
...
+## Writing the update Pipeline job
+
+With CJD now running in our cluster and accessible via HTTPS, we'll next take a look at the Pipeline script that will handle the update of the master. At a high-level, we need our Pipeline to accomplish two major tasks: 
+
+1. build and push our Docker image whenever a change is pushed to the GitHub repository, and
+2. update our Kubernetes resources with the newly built Docker image and any additional changes.
+
+We represent these two procedures as stages within our Pipeline script. 
+
+For the first stage, we will use [kaniko](https://github.com/GoogleContainerTools/kaniko) to build and push our Docker image to a Docker Hub registry. Because we'll be using different agents for each stage, we'll start the Pipeline with `agent none`. Within the first stage, we define our agent using YAML, which specifies the [Google-provided kaniko image](https://gcr.io/kaniko-project/executor:debug) as the container we will use.
```
Comment (kmadel) [inline content/posts/cjd-casc.md]: What about pushing to GCR instead of DockerHub?

File: content/posts/cjd-casc.md [hunk 8]
```diff
@@ -0,0 +1,548 @@
+---
+author:
+  name: "Matt Elgin"
+title: "Self-Updating Jenkins: GitOps for Jenkins Configuration"
+date: 2019-07-01T17:00:00-04:00
```
Comment (kmadel) [inline content/posts/cjd-casc.md]: Update publish date and change `draft` to false once you have reviewed recommended changes. After that, it looks good to me! Great work @mdelgin!

File: content/posts/cjd-casc.md (hunk 1 above)
Comment (mdelgin) [reply content/posts/cjd-casc.md]: Agreed - it was a bit awkwardly phrased. Rewritten for clarity. Thanks for reviewing!

**Pull request merged**